### PR TITLE
OSX mixer

### PR DESCRIPTION
**Pull request by matthewleon**
I have based this off of v0.20.x. I'm not sure what the policy is on this. If you would prefer a version based off of `master`, I can do that as well.

### CODE REVIEW

File: src/output/plugins/OSXOutputPlugin.cxx [hunk 1]
```diff
@@ -103,6 +107,37 @@ OSXOutput::OSXOutput(const ConfigBlock &block)
...
+OSXOutput::GetVolume()
+{
+	AudioUnitParameterValue dvolume;
+	char errormsg[1024];
+
+	OSStatus status = AudioUnitGetParameter(au, kHALOutputParam_Volume,
+			kAudioUnitScope_Global, 0, &dvolume);
+	if (status != noErr) {
+		osx_os_status_to_cstring(status, errormsg, sizeof(errormsg));
+		FormatWarning(osx_output_domain, "unable to get volume: %s", errormsg);
```
Comment (MaxKellermann) [inline src/output/plugins/OSXOutputPlugin.cxx]: Throw an exception instead of logging a warning.

File: src/output/plugins/OSXOutputPlugin.cxx [hunk 2]
```diff
@@ -103,6 +107,37 @@ OSXOutput::OSXOutput(const ConfigBlock &block)
...
+		return 100;
+	}
+	return static_cast<int>(dvolume * dvolume * 100.0);
```
Comment (MaxKellermann) [inline src/output/plugins/OSXOutputPlugin.cxx]: This formula should be explained in a code comment so the next guy understands it.

File: src/output/plugins/OSXOutputPlugin.cxx [hunk 3]
```diff
@@ -103,6 +107,37 @@ OSXOutput::OSXOutput(const ConfigBlock &block)
...
+OSXOutput::SetVolume(unsigned new_volume) {
+	char errormsg[1024];
+	AudioUnitParameterValue scaled_volume =
+		sqrt(static_cast<AudioUnitParameterValue>(new_volume) / 100.0);
+
+	OSStatus status = AudioUnitSetParameter(au, kHALOutputParam_Volume,
+			kAudioUnitScope_Global, 0, scaled_volume, 0);
+	if (status != noErr) {
+		osx_os_status_to_cstring(status, errormsg, sizeof(errormsg));
+		FormatWarning(osx_output_domain, "unable to set new volume %u: %s",
```
Comment (MaxKellermann) [inline src/output/plugins/OSXOutputPlugin.cxx]: Throw exception, see above.

**Comment by matthewleon** [comment]
> please submit a merge into master as soon as I've merged your (corrected) patch in v0.20.x, because I don't have a Mac and I can't test the build.

I am happy to do this. This should be ready for review again.

File: src/output/plugins/OSXOutputPlugin.cxx [hunk 4]
```diff
@@ -103,6 +107,45 @@ OSXOutput::OSXOutput(const ConfigBlock &block)
...
+{
+	AudioUnitParameterValue dvolume;
+	char errormsg[1024];
+
+	OSStatus status = AudioUnitGetParameter(au, kHALOutputParam_Volume,
+			kAudioUnitScope_Global, 0, &dvolume);
+	if (status != noErr) {
+		osx_os_status_to_cstring(status, errormsg, sizeof(errormsg));
+		throw FormatRuntimeError("unable to get volume: %s", errormsg);
+		return 100;
```
Comment (MaxKellermann) [inline src/output/plugins/OSXOutputPlugin.cxx]: Unreachable `return`.

Comment (matthewleon) [inline src/output/plugins/OSXOutputPlugin.cxx]: Thanks. Fixed.

**Pull request merged**